### DISCUSSION TITLE
working-with-the-result-pattern-part-1-replacing-exceptions-as-control-flow/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#436] [Announcements]
# working-with-the-result-pattern-part-1-replacing-exceptions-as-control-flow/

In this post I discuss why you might want to use the result pattern, demonstrate it with a simple example and show a naïve, ugly, solution

https://andrewlock.net/working-with-the-result-pattern-part-1-replacing-exceptions-as-control-flow/

**Comment by bigken** [comment]
the result pattern is common in OpenAPI such as Amazon gateway or Alibaba API.
as the platform, you can tell the caller explictly the error message and the reason.
if you use Exception, it is a nightmare for caller to guess what exception they will get :(

```
{
     "message":"$context.error.messageString",
     "type": "$context.error.responseType",
     "statusCode": "'404'",
     "stage": "$context.stage",
     "resourcePath": "$context.resourcePath",
     "stageVariables.a": "$stageVariables.a"
}```

**Comment by asmorger** [comment]
I am **so** looking forward to this series!  I can't wait to see what you come up with next.  I always enjoy your content, sir!

Fun fact: I have it on good authority that Jeremy has provided that exact feedback to one of my former teams where I implemented that pattern.  Was he referencing my old system? :scream:

I've worked with and implemented several variations of this pattern know over the years. It always goes better with smaller methods, but it never ends up feeling like idiomatic C#.  

I actually had an implementation in my current project, and just recently ripped it out in favor of exceptions.  Does it feel more like orthogonal C#?  Yes.  Are more junior developers able to work with it more easily?  Yes.  Do I still sometimes miss it?  Yes.

**Comment by AndriiStashko** [comment]
I was looking at different implementations of Result pattern some time ago, and the fact there is no standard goto implementation is kind of telling. I personally find it cumbersome. 

What worked for me in the end, though, was `language-ext` library and its construct `Either<L, R>`, where L would be a success value, like a query result, and R would be an error value.

**Reply by mudboy** [reply]
It is kind of funny you say you prefer `Either<L, R>` from `language-ext`, as the example `Result<T>` class shown in this series is essentially an instance of Either with the failure case fixed to `Exception`.

As an aside, it is generally standard practice to use the `R` type in `Either<L, R>` as the success value (it is the "right" value after all) and the `L` for the failure case.

**Comment by otto-gebb** [comment]
> make all that cruft disappear
You mean
```
var x =
  from a in DoSomething1()
  from b in DoSomething2(a)
  select b;
```
?
Nah, pass.

**Comment by modestotech** [comment]
Your articles are the best, Andrew! Love how you show both sides and takes the time to really dive in to subjects in series. After reading Milan's post I thought that it seemed like the "best" solution, but after your reasoning I'm not as convinced. 

Keep yo the good work and looking forward to coming articles!

**Comment by repodn** [comment]
```csharp
public record Error(string Code, string Message);

public class DaNResult<TValue, TError>
{
    public readonly TValue? Value;

    public readonly TError? Error;

    readonly bool _isSuccess;

    private DaNResult(TValue value)
    {
        Value = value;
        Error = default;
        _isSuccess = true;
    }

    private DaNResult(TError error)
    {
        Value = default;
        Error = error;
        _isSuccess = false;
    }

    public static implicit operator DaNResult<TValue, TError>(TValue value) => new DaNResult<TValue, TError>(value);

    public static implicit operator DaNResult<TValue, TError>(TError error) => new DaNResult<TValue, TError>(error);

    public DaNResult<TNextValue, TError> Then<TNextValue>(Func<TValue, DaNResult<TNextValue, TError>> next)
    {
        if (_isSuccess)
        {
            return next(Value!);
        }
        return Error!;
    }
    public DaNResult<TValue, TError> Match(Func<TValue, DaNResult<TValue, TError>> success, Func<TError, DaNResult<TValue, TError>> failure)
    {
        if (_isSuccess)
        {
            return success(Value!);
        }

        return failure(Error!);
    }
}

public record Order (int Price);

public record Invoice(string Text);

public void TestResult()
{
    var result = CreateOrder()
       .Then(order => CreateInvoice(order))
       .Then(invoice => NotifiyUser(invoice));

    result.Match(
        x =>
        {
            Console.WriteLine(x);
            return x;
        },
        error =>
        {
            Console.WriteLine(error.Code);
            return error;
        });
}

DaNResult<Order, Error> CreateOrder()
{
    return new Order(123);
}

DaNResult<Invoice, Error> CreateInvoice(Order order)
{
    return new Invoice("Invoice " + order);
}

```

**Comment by otto-gebb** [comment]
@repodn

```cs
    var result = CreateOrder()
       .Then(order => CreateInvoice(order))
       .Then(invoice => NotifiyUser(invoice));
```

What if `NotifiyUser` needs something from the start of the call chain, e.g. `order`?

**Reply by repodn** [reply]
Instead of partial types, you can pass the whole context to operations and build it up gradually.
Alternatively, you can use more sophisticated monad patterns (Reader, State).

**Reply by otto-gebb** [reply]
Imagine I use `NotifiyUser` in different contexts. Do I create a "god context object" working for each possible call chain? Or do I create many context types, implementing many interfaces like `INotifyUserContext`?

**Comment by osjoberg** [comment]
In my experience and as you demonstrate, using exceptions for control flow is clearly problematic and I would say it is almost always the wrong solution to a problem.

However, I feel that the discussion is not complete if we do not consider how a call site of ProvisionUser() might look like. I am using a made up ASP.Net Core controller as an example here but it could be any calling scenario, really.
```cs
public class UserProvisioningController : Controller
{
	private UserProvisioningService userProvisioningService;

	public ActionResult<UserAccount> ProvisionUser(ExternalLoginInfo info)
	{
		var result = userProvisioningService.ProvisionUser(info);
		if (result.IsSuccess)
		{
			return result.Value;
		}
		
		if (result.Error is ValidationException)
		{
			return BadRequest(result.Exception.Message);
		}
		
		if (result.Error is UnknownTenantException)
		{
			return Unauthorized("Unknown tenant");
		}
		
		if (result.Error is InvalidOperationException)
		{
			return StatusCode(500, "Failed to create provision user request");	
		}
		
		throw new InvalidOperationException();
	}
}
```


As you can see it does not look too bad, except for the handling of the Error. Eh Exception? We are back again to one of the problems with using exceptions as control flow. It is hard to determine from the outside what exceptions that may be returned here and it is not easy to discover if all error scenarios are handled. 

Also, it does not sit right with me to use Exceptions just as a data bearer. Exceptions are meant to be thrown/rarely catched/and handled by a generic handler for unhandled exceptions!

Another thing that stands out to me is that most of the "if" statements flow inside ProvisionUser is repeated inside of the calling code to determine how to handle each error.

I would prefer to skip the usage of exceptions completely and skip the usage of a Result type, and do it like this instead:
```cs
public class UserProvisioningController : Controller
{
	private UserProvisioningService userProvisioningService;
	private CreateUserService createUserService;
	
	public ActionResult<UserAccount> ProvisionUser(ExternalLoginInfo info)
	{
		// TODO: GetClaimValues returns a nullable array which is not optimal. Suggesting to switch to non-nullable array if possible 
		// -or- split it into two methods, one method to check if the information is valid and one method to extract the claim values.
		Claim[]? claims = userProvisioningService.GetClaimValues(info); 
		if (claims is null)
		{
			return BadRequest("The info provided was not valid");
		}

		Claim[] validatedClaims = userProvisioningService.ValidateClaims(claims);
		if (validatedClaims.Length == 0)
		{
			return BadRequest("The claims provided were not valid");
		}

		var tenantId = userProvisioningService.GetTenantId(validatedClaims);
		if (tenantId == Guid.Empty)
		{
			// Possibility to access validatedClaims as needed.
			return Unauthorized("Unknown tenant");
		}
		
		var createRequest = userProvisioningService.CreateProvisionUserRequest(tenantId, validatedClaims);
		if (createRequest == null)
		{
			return StatusCode(500, "Failed to create provision user request");
		}
		
		var identityResult = createUserService.GetOrCreateAccount(createRequest);
		return identityResult;
	}
}
```


This is in my opinion less code, easier to understand and as a minor bonus, likely more performant as well since fewer types are created and collected and fewer branches to execute. 

Where this approach may fall apart if there are many call sites. In those rare circumstances I would probably make a custom type for just that scenario that clearly represents the methods different result states instead of using a Result type which never sits completely right.

**Comment by ivanxpetrov** [comment]
The method should be called Match and not Switch. They are not semantically the same. OneOf makes the distinction and exposes both.
